pkg/strategy/drift/strategy.go

```diff
@@ -0,0 +1,737 @@
+	hiyellow(f, "interval: %s\n", s.Interval)
+	hiyellow(f, "window: %d\n", s.Window)
+	hiyellow(f, "noTrailingStopLoss: %v\n", s.NoTrailingStopLoss)
+	hiyellow(f, "hlVarianceMutiplier: %f\n", s.HighLowVarianceMultiplier)
+	hiyellow(f, "\n")
+}
+
+func (s *Strategy) ID() string {
+	return ID
+}
+
+func (s *Strategy) InstanceID() string {
+	return fmt.Sprintf("%s-%s", ID, s.Symbol)
+}
+
+func (s *Strategy) Subscribe(session *bbgo.ExchangeSession) {
+	session.Subscribe(types.KLineChannel, s.Symbol, types.SubscribeOptions{
+		Interval: s.Interval,
+	})
+	session.Subscribe(types.KLineChannel, s.Symbol, types.SubscribeOptions{
+		Interval: types.Interval1m,
+	})
+
+	if !bbgo.IsBackTesting {
+		session.Subscribe(types.BookTickerChannel, s.Symbol, types.SubscribeOptions{})
+	}
+	s.ExitMethods.SetAndSubscribe(session, s)
+}
+
+func (s *Strategy) CurrentPosition() *types.Position {
+	return s.Position
+}
+
+func (s *Strategy) ClosePosition(ctx context.Context, percentage fixedpoint.Value) error {
+	order := s.Position.NewMarketCloseOrder(percentage)
+	if order == nil {
+		return nil
+	}
+	order.Tag = "close"
+	order.TimeInForce = ""
+	balances := s.Session.GetAccount().Balances()
+	baseBalance := balances[s.Market.BaseCurrency].Available
+	price := s.getLastPrice()
+	if order.Side == types.SideTypeBuy {
+		quoteAmount := balances[s.Market.QuoteCurrency].Available.Div(price)
+		if order.Quantity.Compare(quoteAmount) > 0 {
+			order.Quantity = quoteAmount
+		}
+	} else if order.Side == types.SideTypeSell && order.Quantity.Compare(baseBalance) > 0 {
+		order.Quantity = baseBalance
+	}
+	for {
+		if s.Market.IsDustQuantity(order.Quantity, price) {
+			return nil
+		}
+		_, err := s.GeneralOrderExecutor.SubmitOrders(ctx, *order)
+		if err != nil {
+			order.Quantity = order.Quantity.Mul(fixedpoint.One.Sub(Delta))
+			continue
+		}
+		return nil
+	}
+}
+
+func (s *Strategy) SourceFuncGenerator() SourceFunc {
+	switch strings.ToLower(s.Source) {
+	case "close":
+		return func(kline *types.KLine) fixedpoint.Value { return kline.Close }
+	case "high":
+		return func(kline *types.KLine) fixedpoint.Value { return kline.High }
+	case "low":
+		return func(kline *types.KLine) fixedpoint.Value { return kline.Low }
+	case "hl2":
+		return func(kline *types.KLine) fixedpoint.Value {
+			return kline.High.Add(kline.Low).Div(Two)
+		}
+	case "hlc3":
+		return func(kline *types.KLine) fixedpoint.Value {
+			return kline.High.Add(kline.Low).Add(kline.Close).Div(Three)
+		}
+	case "ohlc4":
+		return func(kline *types.KLine) fixedpoint.Value {
+			return kline.Open.Add(kline.High).Add(kline.Low).Add(kline.Close).Div(Four)
+		}
+	case "open":
+		return func(kline *types.KLine) fixedpoint.Value { return kline.Open }
+	case "":
+		return func(kline *types.KLine) fixedpoint.Value {
+			log.Infof("source not set, use hl2 by default")
+			return kline.High.Add(kline.Low).Div(Two)
+		}
+	default:
+		panic(fmt.Sprintf("Unable to parse: %s", s.Source))
+	}
+}
+
+type DriftMA struct {
```

Comment:
you can move this to the indicator package?

Reply:
this is more of a specific handling to the strategy itself rather than a general indicator.  
won't move it to the indicator package.